### ISSUE TITLE
Clarify usage of aria-haspopup

### ISSUE THREAD

**Issue by scottaohara** [w3c/aria#1024]
Coming out of our discussion of #999

We should review `aria-haspopup` and determine if there should be stronger guidance about what the expectations are for when the attribute should be used, and what sort of variations in behavior should be expected when using the different allowed values.

For quick reference, the current values are:
* false (default)
* true and menu
* listbox
* tree
* grid
* dialog

Specifically regarding the `dialog` value, it was mentioned on the call that when this should be used is ambiguous right now.

As a quick sampling of how this particular value is presently communicated by screen readers when used on a `button` element:

Latest JAWS and NVDA both treat a `button aria-haspopup=dialog` as a menu button, and produce announcements as if the value were `true` or `menu`.

iOS VO announces "shows popup. double tap to activate a picker".  

macOS VO announces it as a "popup button" similarly to a `select` element's announcement.

**Comment by scottaohara** [comment]
To help move discussion on this along [I created a test page](https://scottaohara.github.io/tests/aria-haspopup/) noting how VoiceOver, JAWS and NVDA expose (and by that I mean mostly don't) any of the specific pop-up values.  In most cases, the `aria-haspopup` is indicated by changing the element to append the announcement "pop-up" or "menu" or "submenu".  

Only with JAWS with Chrome and Firefox did the special designations get announced on links.

**Comment by jnurthen** [comment]
NVDA issue for aria-haspopup is https://github.com/nvaccess/nvda/issues/8235

**Comment by carmacleod** [comment]
JAWS issue for aria-haspopup is https://github.com/FreedomScientific/VFO-standards-support/issues/33

**Comment by scottaohara** [comment]
i'm going to move this to 1.3.  I don't think there's time, nor urgency, to tackle this in 1.2.

**Comment by mfairchild365** [comment]
For whatever it is worth, the latest versions of VoiceOver for iOS and MacOS now fully support all values for `aria-haspopup`. Details and test results: https://a11ysupport.io/tests/tech__aria__aria_haspopup_attribute

**Comment by scottaohara** [comment]
that's great to hear.  thanks @mfairchild365

**Comment by joanmarie** [comment]
Just out of curiosity, how important is it to explicitly announce the popup type? If a user is on something with a popup, and activates that popup and winds up in a grid, will the user be completely confused if the screen reader failed to notify them in advance that a grid would appear?

As a sighted user, I don't know in advance what is going to pop up when I activate something with a pop up. I just activate it and see what happens. If a grid appears as a result, my conclusion is that the grid appeared because it's the popup.

To me, tacking on the type seems like it might be unwanted chattiness.

If it's truly important, I can add announcement in Orca. But I was operating under the assumption that this information was mainly of use to the screen reader itself (e.g. an explanation regarding why a grid is claiming to be showing).

**Comment by joanmarie** [comment]
Related question: Combo boxes by default pop up a listbox. Is it considered a screen reader "failure" if it doesn't announce that a combo box opens a list box? Were I the user, my reaction to that announcement might be "of course it does. why are you telling me this?"

**Comment by mfairchild365** [comment]
That's a really good question. There are cases where I think it is important to explicitly announce the popup type.  For example, many ARIA based `<select>` elements have a button with `aria-haspopup='listbox'`. The listbox itself is not displayed until after the button is activated. In these implementations, there is often a visual hint that the button opens a listbox such as a down arrow to the right of the button label. So, `aria-haspopup='listbox'` conveys that same visual information. The APG has an example of this in the [Collapsible Dropdown Listbox Example](https://www.w3.org/TR/wai-aria-practices-1.1/examples/listbox/listbox-collapsible.html).

It may be less important for other values, but I think there are situations where it is very important.

**Comment by carmacleod** [comment]
> Just out of curiosity, how important is it to explicitly announce the popup type?

As long as the screen reader doesn't say "menu button" when `aria-haspopup="dialog"`. :)

**Comment by scottaohara** [comment]
>It may be less important for other values...

and there in lies the source of this issue being created.  :)

**Comment by joanmarie** [comment]
Ok, I just implemented support for it in Orca master: https://gitlab.gnome.org/GNOME/orca/commit/25c64254

Many Orca users (at least those on the Orca mailing list) run Orca master in their production environments (bless their adventurous hearts). Let's see what they say (if anything).

**Comment by carmacleod** [comment]
@mfairchild365 
> there is often a visual hint

Good point. In the case of the APG Collapsible Dropdown Listbox Example, NVDA (for example) says "menu button", which is _almost_ correct, and even has similar keyboard interaction, but doesn't tell the user that they will be "selecting the button's value from a list". (which is kind of an important thing to know, and makes me wonder whether that example should be a combobox instead :) ).

**Comment by carmacleod** [comment]
> Ok, I just implemented support for it in Orca master

Nice! Would love to know what the Orca users say. Please do let us know.

**Comment by joanmarie** [comment]
> That's a really good question. There are cases where I think it is important to explicitly announce the popup type. For example, many ARIA based `<select>` elements have a button with `aria-haspopup='listbox'`. The listbox itself is not displayed until after the button is activated. In these implementations, there is often a visual hint that the button opens a listbox such as a down arrow to the right of the button label. So, `aria-haspopup='listbox'` conveys that same visual information. The APG has an example of this in the [Collapsible Dropdown Listbox Example](https://www.w3.org/TR/wai-aria-practices-1.1/examples/listbox/listbox-collapsible.html).

When an Orca user (using the stable version) encounters that button, Orca says the following:

> Choose an element: Neptunium push button.

When you hear that you know what you need to do conceptually, namely choose an element. In addition, you know what the current choice is, namely Neptunium. If Neptunium ain't your element, what do you do? It's clear that you push the button.

When the same Orca user presses the button, the listbox appears and Orca says the following:

> Expanded
> Choose an element: List with 26 items
> ✓ Neptunium.

Depending on user settings, Orca may also tack on:

> 1 of 26.

That should be enough for the user to know what to do in the list.

I will grant you that failure to present an equivalent of the downward pointing triangle means there is a moment when sighted users have more info than blind users. But I don't personally believe that one moment is significant: You have to choose an element, you're on the thing which lets you choose elements, you activate it and the element-choosing-widget is described so you can choose the element.

But now, Orca master tacks on:

> opens listbox

That makes presentation longer. Speech output is already slower than visual output. And as I describe above, I'm not convinced that longer presentation buys the user anything. :woman_shrugging:

**Comment by mfairchild365** [comment]
Thank you, @joanmarie - that is insightful and you bring up many good points. I agree that in the specific APG example, there is enough information for a user to infer that this button probably opens a list box or some other thing that lets me change the value. I'm less willing to say that holds true for other situations where the name and/or value might not be as clear. Additionally, explicitly conveying the fact that this button opens a listbox might lessen the cognitive load required for someone to understand what they need to do to adjust the value. That's just a hunch, and I don't know that to be true.

I love that you made the change in an effort to collect feedback from end-users. I'm looking forward to hearing the results of that. IMO, that sort of research is one of the best ways to drive decisions like this.

> But I was operating under the assumption that this information was mainly of use to the screen reader itself (e.g. an explanation regarding why a grid is claiming to be showing).

I'd like to better understand this statement. How would this information affect the internals of how a screen reader works? If an author did not provide this attribute, how might that impact the experience of the end-user if it is just used for the screen reader itself?

**Comment by scottaohara** [comment]
>How would this information affect the internals of how a screen reader works? If an author did not provide this attribute, how might that impact the experience of the end-user if it is just used for the screen reader itself?

you mean the attribute, or specific token values of the attribute?

**Comment by mfairchild365** [comment]
Good question. Both.

**Comment by brennanyoung** [comment]
I suggest that the problem is not that NVDA fails to announce the other popup types, but that it specifically mentions **one** of them (arbitrarily chosen by tech history), **but not the others**. 

If the rationale for not announcing the popup role is to reduce chattiness, why does NVDA announce "menu" or "submenu" (which isn't even an aria role) in the first place? Isn't that 'chattiness' too?

Either announce all the enumerated popup roles mentioned in the spec, or announce none at all, and make a briefer, more generic announcement about 'a' popup.

No matter what, **a dialog is very much not a menu**. Can't even get around this mislabelling with `aria-roledescription` in this case.

**Comment by joanmarie** [comment]
> Nice! Would love to know what the Orca users say. Please do let us know.

Not a peep. So they're not bothered by it at least. Though I don't know how much they're encountering it in the wild.

**Comment by sinabahram** [comment]
I think it is critical to talk about behavior in addition to announcement. Please note that aria-haspopup="dialog" results in forms mode being enabled upon invocation in Jaws, and I believe insert mode in NVDA as well. This is highly problematic because the resulting modal dialog, let's say, is then not navigable as a document, independent of its role. This is something that comes up quite a bit on websites we work with developers on where they feel they are doing the right thing, and we tell them, don't at all implement aria-haspopup="dialog" because you will drastically degrade the experience for most screen reader users.

Regarding announcement, it is a good idea to give a priori information to the user when possible, I feel, because it helps set expectations. The metric I always use here is that for an eyes-free user, the cost of undo is almost always higher, often times much higher, than for a mouse or otherwise sighted user (just choosing these two groups for argument, obviously there are many more). Therefore, a minor annoyance can be dismissed with a flick of a wrist, but may take tens of seconds or minutes for a screen reader user to achieve the same result.

As a result, this heads up, I feel, is both useful and honestly necessary. Note that native platforms sometimes have paradigms for this, such as the '...' indicating a new window in MS Windows buttons.

Also, I just wish to say major props to Joanie for implementing this in Orca. Thank you for doing so.

**Comment by JAWS-test** [comment]
> Please note that aria-haspopup="dialog" results in forms mode being enabled upon invocation in Jaws, and I believe insert mode in NVDA as well. 

I have tested this in Chrome and Firefox, with JAWS 2019, JAWS 2020, and NVDA 2019.2.1 Neither for links nor for buttons did using `aria-haspopup=dialog` lead to switching to forms mode. Nor did the use of `role=dialog` switch to forms mode. I could read at any time with the virtual cursor (i.e. with the arrow keys).

**Comment by sinabahram** [comment]
RE forms/insert mode with aria-haspopup, can you please provide a test page, so I can compare results over here? We’re also going to code up some samples over here to make sure we can reproduce the behavior reliably, and I’ll share those as well.

**Comment by JAWS-test** [comment]
@sinabahram : this is my [test page](https://codepen.io/jaws-test/pen/jOEdBBz)

**Comment by sinabahram** [comment]
Thanks for the test page. I just replicated the behavior with Jaws and Chrome, latest versions.

 

I hit that button once, and it turned on forms mode, then turned it off, and then focussed the heading. I then closed the dialog by hitting yes or no, and I then just reproduced the exact same actions; however, this time, the forms mode didn’t toggle off.

**Comment by JAWS-test** [comment]
This is strange, I have no change to forms mode (JAWS 2020, Chrome 79).

**Comment by sinabahram** [comment]
Helps, and maybe it doesn’t, I have the same config over here, same results. Also, this has always been the behavior for as long as I can remember.

**Comment by stes-acc** [comment]
I think it is high time that the WAI-ARIA group **agrees at least on a speech output proposal** for screen readers as part of the definition, for instance **when they replace booleans by string-based enums.** Otherwise it will end up in a support chaos.

I know that ARIA intentionally does NOT define speech output but we can expect that discussions about those WILL appear otherwise as this thread reveals gracefully.

We can and should this clarify beforehand with no ambiguities.

**Comment by jscholes** [comment]
@JAWS-test and @sinabahram, could you confirm your Auto Forms Mode setting value, and test the below assertions?  They apply in both Firefox and Chrome for me.

Activating the button within the provided sample causes forms mode to be activated when focus enters the dialog, if Auto Forms Mode is set to "Manual".  If it is set to "Auto" or "Semi-Auto", this does not happen.

In Firefox, just placing focus on the button causes forms mode to be activated if Auto Forms Mode is set to "Auto" or "Semi-Auto", but not if set to "Manual".

The first assertion seems illogical to me.  If I have Auto Forms Mode completely switched off, I don't expect opening a modal using a button with `aria-haspopup="dialog"` to completely go against that setting.

For the second assertion, the automatic mode change does only apply when Auto Forms Mode is enabled.  But it only applies in a single browser, and is completely unnecessary.  Nobody needs forms mode switched on to press a button.

**Comment by jscholes** [comment]
I should also note that NVDA never turns off browse mode when entering a modal, unless the control to receive focus requires such a change.  A text field, for instance.

**Comment by sinabahram** [comment]
@jscholes, this makes sense. The first thing I do is turn forms mode to manual.

**Comment by JAWS-test** [comment]
I have tested the example at https://codepen.io/jaws-test/pen/jOEdBBz with Chrome. Only in one case JAWS 2019 switches to forms mode: When I open the dialog box by activating the button and have turned off automatic forms mode in the JAWS settings manager (forms mode = manual)

**Comment by jscholes** [comment]
@JAWS-test Thanks for testing.  By "Only in one case", do you mean that 
this only happened once?  Or that it consistently happens with Auto 
Forms Mode set to "Manual".

**Comment by JAWS-test** [comment]
Sorry for the vague language. No, it always happens with the button with aria-haspopup in the dialog when Auto Forms Mode is off. "Only once" means that it does not happen in the following cases:
* when navigating to link and button
* when the link is activated
* with Forms Mode "Auto" or "Semi Auto"

**Comment by stes-acc** [comment]
We should also clarify what value aria-haspopup should have if container has role dialog but the only instance inside is a tree, a menu etc.

These things HAPPEN because developers think of popops being MODAL/NON-MODAL and all available (APG) examples available don't even mention modality of roles tree, menu and so on just saying menu ec. is root container.

The caveat with this is that dialog derives from window whereas menu etc. derives from widget. Window has clearly defined modal/non-modal properties whereas for widget this is just not defined, which causes confusion among devs.

So two questions for clarification:

The ARIA spec is lacking a sentence like "menu role used as container is always modal/non-modal" (pick your preference)

Alternatively, is is OK to "wrap" such popups in dialogs with aria-modal="true/false" and give aria-haspopup the value of the inner content IF THIS IS THE ONLY CONTENT IN THE DIALOG?

**Comment by mcking65** [comment]
I agree with @joanmarie that we should be questioning how much value the aria-haspopup tokens do or even could provide to users. I also agree with @sinabahram that recovering from unintended actions is often more difficult for screen reader users. However, I don't think any of the aria-haspopup tokens can reduce the likelihood of unintended actions that are difficult to undo.

Unfortunately, I played a big part in getting the aria-haspopup tokens into the spec; I drafted the pull requests as part of efforts to resolve combobox issues. As we all know, ARIA 1.1 combobox changes were pretty much a disaster, and it was a lot of work to clean up the mess in ARIA 1.2.	 The aria-haspopup tokens are fall out from that mess, and I see the inconsistent understanding and support as forms of collateral damage, and I regret supporting their addition to the spec. I now see the tokens as solutions looking for a problem.

We added the new haspopup tokens for the combobox role in ARIA 1.1 because there was some concern that screen readers need to know that the combobox would open something other than a listbox. I don't recall the origin of the concern. I have no recollection of how much anyone, including me,  questioned it. We may be able to find more information in meeting minutes from 2016 or so. Regardless, I don't think whatever concerns were expressed at that time have been manifest in practice. So, I don't think screen readers need an attribute that tells them what kind of popup a combobox will open.

In addition, because we were not using the process we have today that requires us to develop authoring practices before adding features to the stable branch, we didn't do a good job of working through all the ramifications of the new values on accessibility APIs and assistive technologies. For instance, we didn't adequately consider the consequences related to the fact that aria-haspopup on a button changes the role of the button to menu button in some accessibility APIS. So, when people started adding haspopup equal dialog on buttons, they got menu buttons that open dialogs -- talk about confusing users!!

If we would have been using today's ARIA development process for ARIA 1.1, then, maybe we would not have changed aria-haspopup.

In the near term, it is not clear to me that there is any real world value to be gained from modifying accessibility APIs and browsers so that assistive technologies can know in advance what kind of popup an element will open. Even if there is some theoretical value, I think we would be better off focusing resources on repairing other ARIA support gaps that do not require accessibility API changes and cannot be avoided by authors. In this case, the solution for authors is simple -- don't use the unsupported token values. Doing so does not create any negative side effects for users.

So, a stopgap corrective course of action could be to invalidate use of any aria-haspopup token except for false, true, and menu unless applying it to a combobox. The combobox exclusion is simply because it is already baked into the pattern, and we don't need any more churn there.

Later, if we come to consensus around additional usage, it would be straightforward to expand support. Fortunately, since we already have some implementation, it would be feasible to research and experiment before charting additional changes, whether they be expanding or removing support. If research shows no benefit, we should consider deprecation. While potentially more frustrating in the near term, it might be the better solution for the overall health of ARIA.

**Comment by JAWS-test** [comment]
I have never understood why aria-haspopup can have different values. While sometimes there is a visual indication that something is opened by an element, I don't know any example where it is visually recognizable in advance if a listbox, tree, grid, menu or dialog is opened. So why does this information need to be perceptible in advance with AT and not visually? Actually, it would be sufficient with AT if this information could be perceived as soon as the additional content is displayed. And this is the case because the content then has the ARIA role listbox, tree, grid, menu or dialog. 
One could even go one step further and abolish aria-haspopup completely and replace it with aria-expanded. aria-expanded also transmits the information that an element can be opened or closed. Visually, there is often no difference between elements marked with aria-expanded and those marked with aria-haspopup. Both elements have a triangle icon to the right of the label. 

> For instance, we didn't adequately consider the consequences related to the fact that aria-haspopup on a button changes the role of the button to menu button in some accessibility APIS

Can this not be changed? (see https://github.com/w3c/core-aam/issues/51)

**Comment by stes-acc** [comment]
> Visually, there is often no difference between elements marked with aria-expanded and those marked with aria-haspopup. 

True, but the technical nature of the popup container matters. Popups are separate nodes not necessarily adjacent  in DOM node sequence and have often different z-order.

**Comment by carmacleod** [comment]
@mcking65 said:
 
> In this case, the solution for authors is simple -- don't use the unsupported token values. Doing so does not create any negative side effects for users.

I just want to mention one tiny use case. We have a select-only combobox that opens a dialog. The spec for [combobox](https://w3c.github.io/aria/#combobox) says:
> Elements with the role combobox have an implicit aria-haspopup value of listbox. If the combobox popup element has a role other than listbox, authors MUST specify a value for aria-haspopup that corresponds to the role of its popup.

Of course, unless we rewrite the spec, validators would complain if the token didn't match (a negative side effect for authors 😄 ).

However, ignoring that little problem for the moment... in this particular use case, I think the token could be useful. A sighted user activating this combobox will not be terribly surprised when they get a dialog. It is visually apparent that the dialog is the popup, because the top of the dialog is aligned with the bottom of the combobox (etc), as well as other visual cues like the dialog is transitioned in when the combobox is opened. So I don't think it's a "cognitive leap" for the sighted user to understand that they dropped down a combobox and got a dialog.

Just guessing, though, that it might be a bit different for a screen reader user. Most of the time, dropping down a combobox gives a listbox. If it's a grid or a tree, well, they're kind of list-like, so no big surprise. But if you're suddenly in a dialog, the reaction might be "Huh? What just happened?". So I'm thinking it might be useful to know ahead of time that the combo has a dialog. Perhaps something as simple as, "Repository combobox, collapsed dialog".

**Comment by JAWS-test** [comment]
@carmacleod: Currently neither NVDA nor JAWS outputs aria-haspopup in any way (tested with Chrome and Firefox at https://codepen.io/jaws-test/pen/abZdbWr)

**Comment by masi** [comment]
IMHO users would benefit from being told that a button opens a dialog. It's a shame that neither JAWS nor NVDA adopted the ARIA 1.1 values.

**Comment by guyhickling** [comment]
>how important is it to explicitly announce the popup type?

I have always considered that aria-haspopup="dialog" would be incredibly useful to screen reader users, if it worked, for two reasons:

a) It lets them know that in order to get back to where they were they have to do something (either look round for a Close button, or press Esc). This is quite different from drop-down content (e.g. accordion style drop downs), where you don't leave the page and you'll eventually get back to the rest of the page content if you just go on navigating long enough.

b) Telling users in advance lets them know what should have happened, in the event that it does not. The greatest problem with dialogs is that most developers do not know how to make them accessible. 

And therein lies the difference between screen reader users and other people. A keyboard-only user can see immediately that a dialog has opened, even if they then find they and the focus have been left out in the cold. 

But a screen reader user has no such information. If focus isn't placed in the dialog, it won't be announced. They are left to hunt around to try to find out what happened when they pressed the button, but won't be able to discover any change on the page. It's just complete silence for them. But around a half of all developers either do not understand that elementary matter of focus, or they're using frameworks that don't manage focus, and testers don't test for it. That is why I consider the dialog announcement MUST be made in advance on the button. (I advise developers to append "Opens dialog" as hidden text to the button or link text since haspopup doesn't work.)

Even when a screen reader user is placed in a dialog, they are often given no announcement about it (if the dialog role isn't added), and often the Close button is unlabelled and/or not in the tabbing order so that clue isn't available either. 

To be fair, there are so many things to get right when creating a dialog - placing focus on it, using the dialog role, adding aria-labelledby to title it, trapping keyboard focus, labeling the Close button, and so on. It's an art in itself and it's easy to see why developers make such a hash of it, or don't even know about it. Whereas adding a single aria-haspopup attribute to the initial button is a very simple thing for developers to learn. It tells screen reader users, in one go, what is supposed to happen, even if they get left out later! 

Just one point, I don't think AT makers should amend the API when they find aria-haspopup, for reasons such as the ones @sinabahram mentions above. They should just announce the impending dialog to the user, and wait to find out what content is inside the dialog before reacting.

And I do think W3C should offer recommended wordings for screen reader announcements for all ARIA attributes. Not statutory wordings - AT manufacturers could still word something differently if they felt it best in a particular case. But a recommended list would stop the current free-for-all that just confuses users. After all, that's what ARIA is for in part, to tell AT what to do, so why not tell them what to say as well? Most screen reader users use more than one reader even if only between their PC and mobile phone. It must be very confusing when switching from one to another to hear such widely differing announcements for the same thing!

About the differences between trees, text dialogs, grids and the rest, I'm less concerned. Once a user lands in a dialog they'll find out what's there. It's when they don't land in it that's the real problem.

Meanwhile it would also be nice to persuade the screen reader manufacturers to put some effort into recognising the haspopup attribute (full marks to Apple there!) instead of just ignoring it. It isn't that difficult, and it would help the users so much. Perhaps the best solution would be to junk haspopup, create a new ARIA attribute for buttons that open a dialog, specify what words should be announced for it, and this time plug it strenuously with the screen reader makers until they capitulate!

**Comment by masi** [comment]
@guyhickling I agree with you. Popup vs dialog is the main distinction. And if screen reader developers have ignored the values for years for no apparent reason let's try it with aria-hasdialog.

**Comment by sinabahram** [comment]
Let's get folks to implement what is there now, not suggest a new attribute that does a subset of what already exists. If they have objections to implementing haspopup, maybe we an spend a fraction of the time it would take to author, debate, build consensus around, and eventually publish a new attribute to just get the AT vendors to implement haspopup or at least document their objections.

**Comment by carmacleod** [comment]
@sinabahram @jnurthen I've created https://github.com/w3c/core-aam/pull/86 to address this by making the core-aam mappings match the ARIA spec. Would love reviews... 2 line fix.

**Comment by mcking65** [comment]
@guyhickling wrote:

> I have always considered that aria-haspopup="dialog" would be incredibly useful to screen reader users, if it worked, for two reasons:
> 
> a) It lets them know that in order to get back to where they were they have to do something (either look round for a Close button, or press Esc). This is quite different from drop-down content (e.g. accordion style drop downs), where you don't leave the page and you'll eventually get back to the rest of the page content if you just go on navigating long enough.

Accordion style drop downs (disclosures) use aria-expanded, not aria-haspopup. They do not move focus or change mode when using JAWS or NVDA.

Other dropdowns, such as menus or listboxes, require JAWS and NVDA users to do something to get out, e.g., escape, enter, tab, you can't read your way out with arrow keys.

>b) Telling users in advance lets them know what should have happened, in the event that it does not.

I appreciate the user-centric intention you are expressing, and empathize with people who struggle to help developers create accessible dialogs.

However, as a user, if the button were to tell me it opens a dialog and then when I activate the button focus doesn't go into a dialog, the attribute on the button that told me about the dialog doesn't necessarily help me. I have no way of knowing if the button was supposed to open a dialog. I'm not going to assume that it should just because it said so. If it didn't open a dialog, I could overlook what the button really did and end up even more frustrated.

I don't advocate support for `aria-haspopup="dialog"` based on its potential to give developers a way to tell users what their buttons should do as opposed to expecting the developer to make the button do what it should do.

**Comment by masi** [comment]
@guyhickling Obviously the announcement for aria-haspopup an what exactly happens after a button activation must match. In ARIA 1.0 there was only the value "true". Jaws and NVDA decided to read out "menu" instead of "pop-up". I assume that this was the reason for the values introduced with ARIA 1.1.

For me the point is not to add new attributes for the sake of it. I'd be content with screen readers stopping to announce "menu" when the button opens a dialog **and** moves the focus. Currently aria-haspop makes only sense in the real world when used with a matching role="menu". But that's not the fault of the spec.

For me the new attribute carries the hope that screen reader developers begin to see the point in telling the users about opening a dialog. Modern web designs has more and more dialogs. OTOH application menus in the style of ARIA role "menu" are rare.

**Comment by zelliott** [comment]
I'm curious what the a11y experts on this thread recommend developers do _for now_ for controls that open dialogs. Should we add `aria-haspopup="dialog"` and live with the poor screen reader support? Should we add `aria-haspopup"true"`? Should we append some informative string `"opens pop up dialog"` to the control's label or description?

[This article](https://html5accessibility.com/stuff/2021/02/02/haspopup-haspoop/) recommends the latter. Other comments on this thread seem to imply using `aria-haspopup` with some value.

**Comment by JAWS-test** [comment]
I would always use aria-haspopup with the correct value. Sometime I guess the output will be correct, hopefully....

As written [above](https://github.com/w3c/aria/issues/1024#issuecomment-707467604), I don't think the wrong output of aria-haspopup is problematic, because a sighted user can't see what will happen to the link or button beforehand either. Only after activating the link or button, the sighted user can see what happened - and so the AT user should also be able to correctly see what happened after activation, e.g. by labeling the dialog with role=dialog, labeling it, giving it focus, etc.

**Comment by jscholes** [comment]
@JAWS-test 

> I don't think the wrong output of `aria-haspopup` is problematic, because a sighted user can't see what will happen at the link or button beforehand either.

This is a bit kcontradictory IMHO.  If a sighted user doesn't know what a button will do when they activate it, why should the screen reader user know what they don't?  More importantly, why should the screen reader user be given information that is likely to be incorrect when a sighted user is given no information at all?

For me the decision comes down to equity.  If a client wants to surface that a button will open a modal for all users, we'll talk about it and find a solution that works for everyone.  Otherwise, knowing that most screen readers will give buttons with `aria-haspopup="dialog"` a role of `menubutton`, it isn't currently worth it.  I'm not a fan of using ARIA to the letter of the spec when I know that screen readers aren't along for the ride.

**Comment by JAWS-test** [comment]
@jscholes 

Since the aria-haspopup bug in buttons exists since many years (and it's not a bug of the AT or the browsers, but of the [specification](https://github.com/w3c/core-aam/issues/51) and thus would be easy to fix by us), I assume that every screenreader user knows it and knows: "Menu Button" is not a button with menu, but a button with aria-haspopup - and I'll be surprised, how the pop-up will look like...

Alternatively, links should be used rather than buttons with aria-haspopup.

**Comment by masi** [comment]
I assumed that aria-haspopup distincts between different targets in ARIA 2.1 because for a screen reader user the context switch is greater than for a sighted user. So this user group gets a hint in advance what to expect when the button gets clicked.

**Comment by jscholes** [comment]
@JAWS-test 

> I assume that every screenreader user knows it and knows: "Menu Button" is not a button with menu, but a button with `aria-haspopup`

We assume very different things.  I assume that most screen reader users don't know `aria-haspopup` exists, and that they listen to what their screen reader is telling them.  When it tells them wrong information, that is a problem.

@masi 

> for a screen reader user the context switch is greater than for a sighted user.

I'm not sure how this should be quantified.  If the target modal is fully accessible, which is far more important than the value of `aria-haspopup` on the triggering button, the context switch may be no more jarring than for sighted users (many of whom will also have strong feelings of love or hate towards the use of a modal in the first place).

**Comment by JAWS-test** [comment]
> I assume that most screen reader users don't know aria-haspopup exists

This does not contradict my statement. I don't need to know that aria-haspopup exists to notice that there are thousands of menu buttons on the web that don't open a menu. However, I'm a big fan of this to fix the problem, but I can't think of more than setting an [issue of the Core AAM specification](https://github.com/w3c/core-aam/issues/51), unfortunately....

**Comment by jscholes** [comment]
@JAWS-test Understood.  Either way, my viewpoint is in the context of advising people how to create accessible and usable experiences.  I don't feel that screen reader users should be forced to adapt to incorrect information, just because they may be able to in the context of less accessible products.  @zelliott I hope this has been in some way helpful.

**Comment by guyhickling** [comment]
zelliott said:
> I'm curious what the a11y experts on this thread recommend developers do for now for controls that open dialogs. 

Whatever is decided here, it will likely take at least 2 or 3 years to feed through into actual compliance by screen readers. 

So when I audit websites, right now, I simply tell the developers to a) not to use aria-haspopup, instead b) append a text of "Opens dialog" to the button label, hidden for sceen readers using the usual CSS. 

There is absolutely no value at all in using the dialog value for aria-haspopup when screen readers don't support it. Recommending its use just misleads developers into thinking it provides appropriate information to blind people when it doesn't. Hidden screen reader text is used for many things, so this is currently the best way to go rather than use an unsupported atribute.

**Comment by JAWS-test** [comment]
Perhaps a distinction should be made between:
* aria-haspopup changes the role and thus partly the screen reader mode for buttons (menu button and change to application mode).
* aria-haspopup is completely ignored
* value in aria-haspopup is not output

I find case 1 (role change) the most problematic. Therefore I would recommend to avoid aria-haspopup on buttons.
Case 2 and case 3 I don't find so problematic for reasons mentioned above, as long as the popup type is correctly perceivable after the popup is opened.

**Comment by scottaohara** [comment]
cc @aleventhal per our discussion this morning regarding `aria-haspopup`.

**Comment by jnurthen** [comment]
https://www.w3.org/2022/01/06-aria-minutes.html#t11

**Comment by scottaohara** [comment]
test cases / results that were sent via email: https://codepen.io/Deafinitive/pen/LYzMJYq

**Comment by scottaohara** [comment]
Discussion largely circled around a proposal to simplify `aria-haspopup` back to being an attribute specifically to be used for creating menu buttons, or to indicate that roles that continue to allow the attribute would be able to open a `menu`.

If this proposal were to be accepted, `aria-haspopup` would be revised to have values of
`true`, `menu` and `false`.

for any deprecated values the expectation would be that they default to 'false' and the use of `aria-haspopup` would no longer be required for roles that presently require them.  

As mentioned, support for the `aria-haspopup` would remain on the elements it is currently allowed on (pending any objections).  

The minor alternative to the above is to also allow a 'generic' value (or some other name) to allow a mechanism to indicate that this particular control can/will 'popup' other content.  Whether that be the token values that are being proposed for removal, or another type of element that is not presently reflected in the current `aria-haspopup` values.

Consensus seemed to prefer the former over the latter proposal, but we would welcome more feedback on which version of simplification is preferred.

To quickly recap
`aria-haspopup` true, menu, false
or
`aria-haspopup` true, menu, generic (or other name), false

**Comment by jscholes** [comment]
Heavily in favour of the former, i.e. true/false/menu only.

**Comment by masi** [comment]
I'd even deprectate "menu". The attribute begins with "has" implying a boolean value. "Upgrading" a boolean value like that is IMHO bad practice and only leads to confusion.

If there is a need for further distinction an aria-popuptype could be added. But I wonder if screend reader developers will adopt that if they did not bother to implement the current non-boolean values of aria-haspopup.

**Comment by jscholes** [comment]
@masi I don't disagree.  However, if web developers have been at least trying to do the right thing by using `"menu"`, I don't think that their mark-up should stop working either.

**Comment by guyhickling** [comment]
> ......we would welcome more feedback on which version of simplification is preferred

@scottaohara If we are voting on those two alternatives, then I go for the "true/menu/false" solution. All other values to default to false or be ignored by AT. 

Because a) that most closely aligns with what most screen readers are doing now, and b) because trying to add any other "generic" value is no more likely to be implemented by the screen reader manufacturers than were the "dialog" and other ARIA 1.1 values.

The only proviso I'd suggest, is to be very firm, very strict, in the ARIA specification, about that bit "all other values to default to false" including the deprecated values. Screen reader manufacturers should be told, firmly, to stop announcing dialogs and other stuff as selectable menus. There can be nothing more confusing for a blind person than to be told they are in a menu and then hear all other kinds of content announced with no sign of a menu!

**Comment by JAWS-test** [comment]
Regarding suggestion 1: If I understand correctly, a button that visually has an arrow and opens a menu would need aria-haspopup=menu or =true. 
A button that also visually has an arrow, but does not open a menu, but a dialog, tree, grid or listbox, must not have aria-haspopup. I would only like this variant if it is made clear at the same time that in these cases the visual indicator should be transmitted via aria-expanded in the future.

The question would be what is special about a menu compared to the other elements. Wouldn't it be sufficient to use aria-expanded for all buttons and completely dispense with aria-haspopup?

If aria-haspopup is to be for menus only: In the long term, the aria-haspopup=menu could then also be abolished, so that there would then only be true and false.

**Comment by stes-acc** [comment]
I would not deprecate any of them **just because of temporary pathetic support in screen readers.** 

I also won't accept arguments like "people are not used to that" or "it is used wrong or even dangerous" because these are no arguments at all. If taken srious, e-bikes would never succeded. It's all about the BENEFITS new things bring in the long run that causes people to jump on the new wagons and go west.

If there are logical implications inside the ARIA model to change values, or if their string values are to be revised because of lack of clarity, use cases etc. that's annother thing.
 
I just can say that there ARE use cases for "dialog" and "grid" and I will defend them any time.

**Comment by masi** [comment]
> The question would be what is special about a menu compared to the other elements. Wouldn't it be sufficient to use aria-expanded for all buttons and completely dispense with aria-haspopup?

As I read it aria-expanded makes a portion of the DOM accessible that at it ends will not move the focus back to the element itself.
aria-haspopup signals a new portion of the DOM that will trap the focus and must explicitly left with a user command like pressing escape.

**Comment by jscholes** [comment]
@JAWS-test `aria-expanded` indicates state, `aria-haspopup` conveys intent/impact.  As @masi pointed out, `aria-haspopup` is often intended to indicate that a control will change the user's context to a modal component (note that modal does not equal dialog).  It does not feel reasonable for the purpose of `aria-expanded` to be overloaded in this way, nor should its state be relevant to the user while in a modal context.

**Comment by scottaohara** [comment]
@masi, deprecating `menu` is not off the table.  But doing so would be more to indicate to authors to just use the `true` value instead, but it would still functionally serve the same purpose.  Where as the deprecation/removal of the other values would be equivalent to authors specifying `aria-haspopup=false` or simply not using the attribute at all.

Regarding:
>As I read it aria-expanded makes a portion of the DOM accessible that at it ends will not move the focus back to the element itself.
aria-haspopup signals a new portion of the DOM that will trap the focus and must explicitly left with a user command like pressing escape.

Sort of?  Using `aria-haspopup` does not explicitly indicate that focus will be trapped.  The element that popups is what should set the expectation for behavior. In reality, it is often that popups which are not modal dialogs **do not** completely trap focus.  Rather, <kbd>Tab</kbd> key can be used to navigate away from the popup and it will auto-dismiss. menus and listboxes, for example.

**Comment by scottaohara** [comment]
Further thoughts I've been having on this.

if `aria-haspopup` is reduced to `true`, `menu` and `false` - essentially to allow a button to have it's role changed to a 'menu button', then I wonder what use case does `aria-haspopup` have on any other element?   The mappings for those elements don't change.  So what would make a 'menu' special to note as being a popup for those elements, but other popup types would not need to be indicated?

If `aria-haspopup` wouldn't allow for a generic value to indicate that 'something' besides a menu would popup, then I question why not deprecate the attribute all together, and create a new `menubutton` role for authors to migrate their `button aria-haspopup="true | menu"` elements to.

If it is determined that the control that invokes a popup of X type does not need to provide a programmatic cue that it will do so, then I think there is value in providing a mechanism to indicate that the UI that has popped up is in fact a popup.  For example, using `aria-expanded` alone does would not imply that a list of links has been displayed as a popup, and that tabbing away from it will auto-dismiss the that content.  Even if putting that list of links into a non-modal dialog, which would at least indicate that the element is not part of the underlying page, it wouldn't necessarily be expected that tabbing out of that dialog would close it.

**Comment by guyhickling** [comment]
I wouldn't like to see the "menu" value deprecated, because it works very well currently. Menus with that are announced as menus by most screen readers, so that's good. 

The aria-haspopup="true" works not quite so well, because some screen readers announce it as "Menu" while others just announce "Popup". But it is still handled fairly reasonably.

The reason for deprecating all the other values is because if we don't, then some developers will continue to use them, in the fond belief they are doing something useful for screen reader users when they aren't. Grief, I did that myself - for two years I kept telling one client after another to use the dialog value, on the naive assumption it was only a matter of time before the screen readers came on board. But they didn't, so now I just tell clients to add screen reader text to say the same thing. 

Sometimes the W3C amends specs to accept the status quo. There is advantage in doing that.

**Comment by masi** [comment]
@scottaohara
> > aria-haspopup signals a new portion of the DOM that will trap the focus and must explicitly left with a user command like pressing escape.
> 
> Sort of? Using `aria-haspopup` does not explicitly indicate that focus will be trapped. The element that popups is what should set the expectation for behavior. In reality, it is often that popups which are not modal dialogs **do not** completely trap focus. Rather, Tab key can be used to navigate away from the popup and it will auto-dismiss. menus and listboxes, for example.

Ah yes, sorry, forgot the auto-closing feature. But then again this is not what I would expect from anything opened with aria-expanded, ie that it shrinks (closed) automatically. Or is it okay to that as well?

**Comment by masi** [comment]
> If `aria-haspopup` wouldn't allow for a generic value to indicate that 'something' besides a menu would popup, then I question why not deprecate the attribute all together, and create a new `menubutton` role for authors to migrate their `button aria-haspopup="true | menu"` elements to.

Has this not been suggested already?

BTW, does anyone know why the attributes "new" values have never been adopted? Are there no representatives of the vendors or developers here?

**Comment by scottaohara** [comment]
@guyhickling `true` and `menu` are supposed to be the same thing.  If you are finding that they are treated differently, then that's a bug.

@masi the comments I posted here today are ideas that have not been discussed yet.  

Regarding the ARIA 1.1 values - it's not exactly true that they haven't been adopted.  The major problem is with the `button` element because `aria-haspopup` is mapped to a "menu button" role, e.g., MSAA/IA2 `ROLE_SYSTEM_BUTTONMENU`.

It's not a 100% AT has dropped the ball on these values.  Browsers have not consistently changed the mappings either.  But these values do work sporadically on other elements.

Again though... it's sort of "what's the point?"  Why are these very specific things allowed to be popups, when in reality there is a larger variety of UI that "pops up".

**Comment by JAWS-test** [comment]
>  The major problem is with the button element because aria-haspopup is mapped to a "menu button" role, e.g., MSAA/IA2 ROLE_SYSTEM_BUTTONMENU

My proposal has always been to abolish precisely this: https://github.com/w3c/core-aam/issues/51

**Comment by mitchellevan** [comment]
As a sighted web developer, I used to have high hopes that `aria-haspopup` would gain support and express useful information about the most common patterns of dynamic content. Banner navigation fly-outs and toggle tips typically provide contextual visual cues that they will display content within the page, and the visual cues help set users' expectations for distinct interaction patterns. Controls which trigger a modal dialog usually do not convey that fact visually before a user activates them, but nevertheless I wanted to give screen reader users a heads-up about a potentially disorienting dialog.

Things haven't turn out as I hoped. It's been a long time since I last coded or recommended `aria-haspopup`. These days a combination of `aria-expanded`, semantic roles of triggers and dynamic content, and focus management (where appropriate) is deemed adequate.

However, I still wish we could use `aria-haspopup` for certain common types of [disclosures](https://www.w3.org/TR/wai-aria-practices-1.1/#disclosure). Some disclosures display their content inline (like CSS `position:static`), while others display their content in a layer on top of all other content (like CSS `position:relative`). For almost all inline disclosures, users can hide them only by re-activating the trigger. For disclosures on top of other content, users can hide them using other mechanisms, such as pressing ESC or tapping elsewhere on a touchscreen. The button role and `aria-expanded` attribute are the same for both. In the absence of contextual cues, screen reader users are not informed of the different hiding mechanisms.

So between the two choices @scottaohara listed from the Working Group, I vote for proposal 2. A new "generic" value would give screen reader users the possibility of distinguishing between those two types of disclosure interactions with less trial-and-error.

Straw man: Deprecate all except "true" and "false". Allow `aria-haspopup="true"` for all kinds of dynamic content that appears in front of other content. This would be refreshingly simple for web developers. — Argument against: the "true" value is too firmly established as the same as the "menu" value, and only [desktop-like menus](https://www.w3.org/TR/wai-aria-practices-1.1/#menu) should be conveyed as menus. (If I'm wrong, and there's actually a realistic path to this end state, then I would be all for it.)

Another straw man: Get rid of `aria-haspopup` entirely. Rely instead on a role for generic disclosed content, such as non-modal dialog or maybe a new "popup" role. Again, from a developer's perspective this approach is appealing, as they would be parallel to other roles of disclosed content (e.g., dialog). — Argument against: I want to fix `aria-haspopup`, not get rid of it, because it meets a user need. As @sinabahram said:

> it is a good idea to give a priori information to the user... for an eyes-free user, the cost of undo is almost always higher

**Comment by jscholes** [comment]
@mitchellevan A disclosure that causes content to obscure other content, and must be dismissed before that other content becomes usable again, sounds like a small modal dialog.  Or, at the very least, something that should be given a role of `dialog` to indicate that it has defined boundaries, and wants the user's attention.  I don't agree that `aria-haspopup` should ever serve the purpose of conveying such UI techniques, because there is more at play than just the triggering button.

**Comment by scottaohara** [comment]
@jscholes a bit of a re-frame from that scenario.

There are plenty of instances of content that 'pops up' that don't neatly fit into a dialog (modal or modeless) because, like a popup `menu` or `listbox`, when someone tabs away from the container the popup dismisses.  

tbh, i'm far less concerned with whether `aria-haspopup` is used to convey that a user is going to interact with such UI, as much as I think it's worth exposing the fact that the content which has popped up has this type of behavior.  

that could be a 'popup' qualifier that is appended to the primary content container that a user has found themselves in, when invoked.  -  or JAWS help text like content that a user could turn on or off as they wish to indicate that if someone were to navigate away from the popup, it would close.  Or it could be an announcement is made when such content 'collapses' to indicate "that thing you were just in, you have to open it again"

**Comment by scottaohara** [comment]
next steps, @mcking65 and I need to work on a suggested PR.

another issue needs to be created (by me) regarding needing to inform users that a "popup" will have light dismiss behavior.

**Comment by guyhickling** [comment]
> and I need to work on a suggested PR

So have you decided what the change will be? What will the PR advise and contain?

**Comment by jnurthen** [comment]
@guyhickling you can read the discussion at https://www.w3.org/2022/02/03-aria-minutes.html#t07

**Comment by masi** [comment]
Those minutes are incredibly hard to read. Only thing I could make out that there is no consenus at all.

Issues are IIUC mapping to the accessibility API of the OS, backwards compatibility with the spec, add properties or roles to a new version of the spec,  implementation problems of current screen readers, lack of definition on what a pop-up actually is * and if non-sighted users need additional information.

* I assume that even if ARIA would define pop-up the definition will not match the average developers notion of a pop-up. I for example don't use the word drop-downs of a html select element.

**Comment by scottaohara** [comment]
I used the term 'suggested' purposefully when describing the todo to create the PR.

the point of working on a suggested PR is to help get consensus.  Matt and I have ideas of where this should go, but they're not 100% aligned.  We're going to work together to align them as best we can and then make the suggestion so that we have something concrete to talk about with the group.  otherwise, we may all just continue to circle around the solution.

**Comment by masi** [comment]
> Matt and I have ideas of where this should go, but they're not 100% aligned. We're going to work together to align them as best we can and then make the suggestion so that we have something concrete to talk about with the group. otherwise, we may all just continue to circle around the solution.

Thank you for going into all that troubles.

**Comment by daemoncron** [comment]
Just chiming in here, in 2022, this attribute has been around for a while, and still every time we see a screen reader user encountering `aria-haspopup="dialog"` in a user test, they are confused by it - generally thinking it is a menu (even when "dialog" is announced) because of the history of "popup" being a menu.

I think I'm going to advise we move away from using this on buttons that launch modals in favor of screen reader only text or `aria-label`s (kind of like "(opens in new window)" we could just use "(launches modal dialog)"), unless there's a better idea out there.

**Comment by sidnc86** [comment]
For ages, as a screen reader user on Windows, a major queue of notifying `aria-haspopup="dialog"` has been use of ellipsis. For example, in MS Word, if one had to bring up a paragraph dialog, the option would show up as "Paragraph...". With Punctuation Level setting of most screen reader is set to either "some" or "most" by default, amny screen reader users won't fiddle with settings unless they are advanced users. The reason I bring up Punctuation Level setting is because the ellipsis indication is translated into "," like pause. Does the pause really help? This is going to be quite subjective. But I thought to put the ellipsis usage in context because although not through direct roles, there has been ways to communicate that something pops up a dialog even at that non-web user interface.
As  programmer, I thought aria-haspopup="dialog" was intended to add this ellipsis value to screen reader user's life. It is sad that screen reader vendors haven't yet implemented aria-haspopup="dialog".
I am not a big fan of using sr-only off-screen text to convey role, property or state information, that should be confined to giving out screen reader specific instructions.

**Comment by giacomo-petri** [comment]
Hello everyone, 

thank you for raising this topic. It is an interesting one with a lot of comments and a very long discussion (I would have noticed it before).
After lot of years discussing the aria-haspopup topic with my team, I'm still struggling with it.

As always there are a lot of insightful hints, comments and thoughts.

----

From @sinabahram:

> Regarding announcement, it is a good idea to give a priori information to the user when possible, I feel, because it helps set expectations. The metric I always use here is that for an eyes-free user, the cost of undo is almost always higher, often times much higher, than for a mouse or otherwise sighted user (just choosing these two groups for argument, obviously there are many more).

+1, I totally agree with this.

----

From @mcking65  

> We added the new haspopup tokens for the combobox role in ARIA 1.1 because there was some concern that screen readers need to know that the combobox would open something other than a listbox. I don't recall the origin of the concern. I have no recollection of how much anyone, including me, questioned it. We may be able to find more information in meeting minutes from 2016 or so. Regardless, I don't think whatever concerns were expressed at that time have been manifest in practice. So, I don't think screen readers need an attribute that tells them what kind of popup a combobox will open.

This is good to know but at the same time, in my opinion, it's not intuitive from specs. The "dialog" value seemed a bit off-topic considering the entire list of values and that's also why I've always thought about it as a useful hint to inform AT (screen reader) users about something (dialog) that will be visually displayed on screen after interacting with the element. In my mind, this interpretation was also supported by the list of "Used in Roles", which contains for example the link role (which usually is not used to display "menu, listbox, tree or grid" elements). That said, with the combobox information and assumption, the list of values makes more sense.

----

From JAWS-test
> I have never understood why aria-haspopup can have different values. While sometimes there is a visual indication that something is opened by an element, I don't know any example where it is visually recognizable in advance if a listbox, tree, grid, menu or dialog is opened. So why does this information need to be perceptible in advance with AT and not visually?

And answer from guyhickling
> a) It lets them know that in order to get back to where they were they have to do something (either look round for a Close button, or press Esc). This is quite different from drop-down content (e.g. accordion style drop downs), where you don't leave the page and you'll eventually get back to the rest of the page content if you just go on navigating long enough.

b) Telling users in advance lets them know what should have happened, in the event that it does not. The greatest problem with dialogs is that most developers do not know how to make them accessible.

As also stated previously by @sinabahram, in my opinion this is a key point. Different experience requires different cues. I'll expand this in my final thoughts.

----

From @zelliott 
> I'm curious what the a11y experts on this thread recommend developers do for now for controls that open dialogs. Should we add aria-haspopup="dialog" and live with the poor screen reader support? Should we add aria-haspopup"true"? Should we append some informative string "opens pop up dialog" to the control's label or description?

That's a very good question.
If the intent was to use the aria-haspopup attribute in combination with the combobox role I would say that, even if it's not properly supported by AT (only VO is properly informing users about correct value) in terms of user experience it might still work.
But considering that the aria-haspopup can be used also with other roles and, per my initial assumption, that it could be used as a hint useful to inform users about something (dialog) that is visually displayed on screen, with the current AT support (which didn't change during the past years) it can confuse users.
Do we have a final answer for this?

----

From @scottaohara

> Discussion largely circled around a proposal to simplify aria-haspopup back to being an attribute specifically to be used for creating menu buttons, or to indicate that roles that would continue to allow the attribute (if any?) would allow for opening a menu.
> 
> If this proposal were to be accepted, aria-haspopup would be revised to have values of:
> true, menu and false. menu remaining a valid value as its a synonym for true in this context.
> 
> for the remaining deprecated values, they would instead map to 'false. The use of aria-haspopup would no longer be required for roles that presently require them (e.g.,combobox`).

This might be an interesting proposal, but what the disability community - especially VO users which might be more familiar with something that currently works - thinks about it? Do they feel comfortable from no longer receiving hints about buttons that trigger dialogs? Are them somehow impacted from removing this support?

----

From @mitchellevan 

> As a sighted web developer, I used to have high hopes that aria-haspopup would gain support and express useful information about the most common patterns of dynamic content. Banner navigation fly-outs and toggle tips typically provide contextual visual cues that they will display content within the page, and the visual cues help set users' expectations for distinct interaction patterns. Controls which trigger a modal dialog usually do not convey that fact visually before a user activates them, but nevertheless I wanted to give screen reader users a heads-up about a potentially disorienting dialog.

+1, I think this is what AT users expect.

----

From @scottaohara

> There are plenty of instances of content that 'pops up' that don't neatly fit into a dialog (modal or modeless) because, like a popup menu or listbox, when someone tabs away from the container the popup dismisses.
> 
> tbh, i'm far less concerned with whether aria-haspopup is used to convey that a user is going to interact with such UI, as much as I think it's worth exposing the fact that the content which has popped up has this type of behavior.

As you said, excluding combobox elements, there are still many instances of content that pops up; fly-outs, dialogs (modal and non-modal), tooltips (meant as generic element that appears on focus or over - sometime might be just a description but some other might contain operable controls or multiple details, e.g. password requirements), just to name a few.
And many times the aria-expanded attribute is not enough to convey what will happen (as @stes-acc mentioned, the aria-expanded is a state attribute, whereas aria-has-popup isn't).

That said, I'm not convinced that aria-haspopup="dialog" implicitly means that tabbing away from the component results in dismissing the popup. In the document there is no definition of "popup"; if this is the intent might be worth adding it.

Per aria-haspopup property definition

> Indicates the availability and type of interactive popup element, such as menu or dialog, that can be triggered by an element.
> 
> A popup element usually appears as a block of content that is on top of other content. Authors MUST ensure that the role of the element that serves as the container for the popup content is menu, listbox, tree, grid, or dialog, and that the value of aria-haspopup matches the role of the popup container.

Based on that, it might be used on all scenarios I've listed (fly-outs, modal and non-modal dialogs, etc.), which might not be dismissed automatically after tabbing away.

----

From @daemoncron 

> Just chiming in here, in 2022, this attribute has been around for a while, and still every time we see a screen reader user encountering aria-haspopup="dialog" in a user test, they are confused by it - generally thinking it is a menu (even when "dialog" is announced) because of the history of "popup" being a menu.

This is interesting; I understand and agree with that if using JAWS or NVDA. That said I experienced the opposite with real VO user testing activities.

----

My final thoughts:
I've always approached accessibility assuming that all users with different abilities must all be able to accomplish their task, but the experience, the "perceived journey", the modalities of interaction, the timing spent, etc., might be different. A website is compliant with WCAG when users are able to perform website tasks; a website is usable/accessible when users are satisfied about their navigation (and the efficiency of their navigation).
To better clarify my point and taking this topic as example, even if there is no visual indicator that an element will trigger the presentation of a dialog (dialog intended as modal dialog, that prevents users navigating outside also with the tab key), for screen reader users might be easier if they are able to predict what will happen interacting with the element; for example just being aware that pressing the escape key they will be able to dismiss it and return to the previous content is a good hint (they may try to use the back browser functionality instead). Also a low vision user (e.g. tunnel vision) that is helped by a screen reader might benefit about the announcement of the popup, if the popup is not presented next to the button that triggered it, knowing they might need to search for it around the page. Of course a sighted user, after interacting with the popup-button, is immediately able to recognise the widget, is able to see the "inert" background page with probably an opacity overlay between the background page and the modal, etc. 
Experiences are different, perceptions are different, but allowing screen reader users predicting the dialog behaviour might be something that can be compared to the immediate feedback perceivable visually by a sighted user. 

Similarly, when a link is opening a new window (target="_blank"), it's recommended to inform AT users about that (even if there are no visual cue), as for a blind user it might require more time to understand what happened and to return to the previous page (a bit off topic but I still don't understand why there is no way to map the target attribute to make AT announcing this detail). Is it not similar to a button element that opens a dialog? Maybe less intrusive but still something not immediate.

Consequently, I'm not sure that providing an AT hint only when it's visually available is fair enough, and I don't think that moving back to just three values ("true", "menu", "false") is ok, unless there is another way to convey that detail.

**Comment by spectranaut** [comment]
here are some comments from scott: https://gist.github.com/scottaohara/7ef3f2882c911824ac2da35294f6beb9

**Comment by cletusw** [comment]
Haven't tested personally, but it looks like latest JAWS (https://github.com/FreedomScientific/standards-support/issues/33) and NVDA (https://github.com/nvaccess/nvda/pull/14709 ) have fixed this now.

Table (not updated for NVDA yet, maybe it hasn't shipped a version with the fix): https://html5accessibility.com/stuff/2023/06/20/aria-haspopup-less-is-more/

**Comment by jnurthen** [comment]
Has this been resolved or is this still an issue? This issue is really long - if something still needs to be done could someone please create a PR for what needs to be done to resolve.

**Comment by jnurthen** [comment]
Based on https://html5accessibility.com/stuff/2023/06/20/aria-haspopup-less-is-more/ can we now close this?

**Comment by mitchellevan** [comment]
> @joanmarie
> 
> > Just out of curiosity, how important is it to explicitly announce the popup type?
> 
> As long as the screen reader doesn't say "menu button" when `aria-haspopup="dialog"`. :)

Because I agree with the above, and because this minimum support or better is now in latest versions of screen readers with widespread use, I say this issue should be closed without a change in ARIA.

A content creator might consider it important to support older screen reader versions, and therefore continue avoiding the newer values such as `aria-haspopup="dialog"`. But it's safe to say back-compatibility concerns for `aria-haspopup` are now in the same range as for other ARIA features.

**Comment by scottaohara** [comment]
i think for now this can be closed.  still things to consider with the actual practicality of this attribute... but those can be new issues.

**Comment by LaurenceRLewis** [comment]
I have always thought the following part of the spec confusing, 

> "Authors MUST ensure the popup element associated with a combobox has a role of listbox, tree, grid, or dialog." 

because authors MUST is followed by;

> "Elements with the role combobox have an implicit aria-haspopup value of listbox. If the combobox popup element has a role other than listbox, authors MUST specify a value for aria-haspopup that corresponds to the role of its popup."

These two paragraphs should be rewritten as;

"Elements with the role combobox have an implicit aria-haspopup value of listbox. If the combobox popup element has a role other than listbox, authors MUST specify a value, tree, grid, or dialog, for aria-haspopup that corresponds to the role of its popup."

**Comment by mitchellevan** [comment]
> I have always thought the following part of the spec confusing,
> 
> > "Authors MUST ensure the popup element associated with a combobox has a role of listbox, tree, grid, or dialog."
> 
> because authors MUST is followed by;
> 
> > "Elements with the role combobox have an implicit aria-haspopup value of listbox. If the combobox popup element has a role other than listbox, authors MUST specify a value for aria-haspopup that corresponds to the role of its popup."

This spec language is technically accurate as-is, because "implicit" does not mean informally implied or hinted at. It means the property has a specific value, programmatically determined from the element's role, as described in [Implicit Value for Role](https://www.w3.org/TR/wai-aria/#implictValueForRole).

Is it as clear as it could be? Maybe not... So I'd suggest requesting that clarification as a new issue, or go straight to a PR. This should be separate from #1024, so #1024 can be finally closed.

> Which brings me to wonder how relegating aria-haspopup to true/menu only will affect role combobox.

I don't foresee that happening. Such a change in spec was being considered in past years, when some major screen readers were either not announcing information for non-`listbox` values of `aria-haspopup` or were announcing information inconsistent with the value. However, now that major screen readers are announcing information consistent with the existing spec, there is no longer a need for such a change in spec, and #1024 should be closed without a change.

**Comment by scottaohara** [comment]
this issue _is_ closed.
